doc/engine.md

```diff
@@ -0,0 +1,215 @@
+# Guidelines for engine contributions
+
+This document is meant to clarify and codify the standard
+by which engine contributions may be judged in relation
+to the mission statement of the Minetest project.
+
+Minetest is a game engine and a platform for authoring,
+distributing and running 3D voxel games, single and multiplayer.
+
+## 1. The server model and API design
+
+Minetest employs a client-server model at all times.
+Even in singleplayer, a logical server
+is created and the client is made to connect to it.
+
+At the moment, a game is not allowed to send any
+code for the client to execute, so all game code
+is effectively executed only on the server.
```

Comment:
Code could be executed by client-installed CSMs using the mod channels API. This is a very niche feature, but possible in general. "Not allowed" is IMO not the right expression. I'd suggest the following:

"At the moment, a game is entirely executed on the server and offloading mod functionality to clients is not in an usable state."

Reply:
Using existing APIs in a vanilla install it is indeed not allowed. This loophole doesn't really apply since you need to manually install a gaping security hole into your copy, by utilizing an experimental, unsupported feature. I almost think it would be harmful to acknowledge that possibility.

From the perspective of a well meaning modder it's just impossible. And this is not really a technical document, it doesn't need to cover everything.